[package] resolve `__import__` calls on export

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55153 [package] resolve `__import__` calls on export**

There's no logic on the export side to capture dependencies created by` __import__(...)` calls. This PR adds the logic to capture the dependencies on PackageExport's find_files_source_depends_on calls 

Differential Revision: [D27504536](https://our.internmc.facebook.com/intern/diff/D27504536)

Commit:
[package] resolve `__import__` calls on export

[ghstack-poisoned]

suo commented on torch/package/find_file_dependencies.py:
```diff
@@ -41,5 +41,48 @@ def visit_ImportFrom(self, node):
+        # __import__ calls aren't routed to the visit_Import/From nodes
+        if hasattr(node.func, "id") and node.func.id == "__import__":
+            name = node.args[0].value
+            fromlist = None
```
nit: might as well initialize fromlist to an empty list here

suo commented on test/package/test_save_load.py:
```diff
@@ -68,6 +68,19 @@ def test_save_module(self):
+    @skipIf(IS_FBCODE or IS_SANDCASTLE, "Tests that use temporary files are disabled in fbcode")
+    def test_dunder_imports(self):
+        filename = self.temp()
+        with PackageExporter(filename, verbose=True) as he:
```
We should clean up the body of this test: 
- verbose=False
- no print
- We should have assertions that the `__imported__`'d modules are accessible from the resulting package.

Lilyjjo replied:
For a module to be accessible, do you mean that somethiing along the lines of `hi.import_module("subpackage_2") should work?

Commit:
Update on "[package] resolve `__import__` calls on export"


There's no logic on the export side to capture dependencies created by` __import__(...)` calls. This PR adds the logic to capture the dependencies on PackageExport's find_files_source_depends_on calls 

Differential Revision: [D27504536](https://our.internmc.facebook.com/intern/diff/D27504536)

[ghstack-poisoned]

Commit:
Update on "[package] resolve `__import__` calls on export"


There's no logic on the export side to capture dependencies created by` __import__(...)` calls. This PR adds the logic to capture the dependencies on PackageExport's find_files_source_depends_on calls 

Differential Revision: [D27504536](https://our.internmc.facebook.com/intern/diff/D27504536)

[ghstack-poisoned]

Lilyjjo commented on torch/package/package_importer.py:
```diff
@@ -402,6 +402,9 @@ def _handle_fromlist(self, module, fromlist, *, recursive=False):
                         self._handle_fromlist(module, module.__all__, recursive=True)
                 elif not hasattr(module, x):
                     from_name = "{}.{}".format(module_name, x)
+                    # handle case where from_name is ''
```
I'm not sure why this change needs to be made-- I'm gonna guess it's because not all the parts that flow into/out of this function are the same as cpython's. The cpython's eventual logic calls `_find_and_load` which is different from ours, it could handle it there maybe. It's valid python to have the `fromlist=[""]` which is what triggers this case. It's not related to how the export's dunder import is written because the error is on the importing side not being able to resolve a package with a dot at the end:
```
__import__("subpackage_0.subsubpackage_0", globals(), fromlist=[""], level=1)
## Resulting error:
ModuleNotFoundError: No module named "package_b.subpackage_0.subsubpackage_0." in self-contained archive "<binary>" and the module is also not in the list of allowed external modules: ['']
```
The dot is added when the fromlist is parsed on line 404

suo replied:
I figured this out, the `ModuleNotFoundError` should be suppressed by line 412 below, but it relies on a check against `exc.name`. 

Unfortunately, our implementation of the loader does not correctly populate the `name` attribute when raising a `ModuleNotFoundError`. 

The fix is to add `name=name` to the `ModuleNotFoundError` constructor on line 289.

Commit:
Update on "[package] resolve `__import__` calls on export"


There's no logic on the export side to capture dependencies created by` __import__(...)` calls. This PR adds the logic to capture the dependencies on PackageExport's find_files_source_depends_on calls 

Differential Revision: [D27504536](https://our.internmc.facebook.com/intern/diff/D27504536)

[ghstack-poisoned]

suo commented on test/package/package_b/__init__.py:
```diff
@@ -0,0 +1,15 @@
+__import__("subpackage_1", globals(), fromlist=["PackageBSubpackage1Object_0"], level=1)
+__import__("package_a", fromlist=[])
```
when doing automatic test case discovery, `unittest` will execute all `__init__.py` files to try to find more tests to run. This import will fail because `package_a` is not in the module search path at that point.

I suggest changing this to importing a subpackage, which will probably just work (you can run `python test/test_package.py` to check).

suo reviewed (changes requested):
two minor cleanups and I think we are good!

Commit:
Update on "[package] resolve `__import__` calls on export"


There's no logic on the export side to capture dependencies created by` __import__(...)` calls. This PR adds the logic to capture the dependencies on PackageExport's find_files_source_depends_on calls 

Differential Revision: [D27504536](https://our.internmc.facebook.com/intern/diff/D27504536)

[ghstack-poisoned]

Commit:
Update on "[package] resolve `__import__` calls on export"


There's no logic on the export side to capture dependencies created by` __import__(...)` calls. This PR adds the logic to capture the dependencies on PackageExport's find_files_source_depends_on calls 

Differential Revision: [D27504536](https://our.internmc.facebook.com/intern/diff/D27504536)

[ghstack-poisoned]
